[IP-1] Working expense module for 1.5.12

chrispytorius opened the pull request:
Sorry i messed up my first pull request, this one is has the correct files. Git is new to me :/

Pull Request Checklist

  * [x ] My code follows the code formatting guidelines.
  * [ x] I have an issue ID for this pull request. 
  * [ x] I selected the corresponding branch.
  * [ x] I have rebased my changes on top of the corresponding branch.
  
Issue Type (Please check one or more)

  * [ ] Bugfix
  * [ ] Improvement of an existing Feature 
  * [ x] New Feature

pradeshc commented:
@chrispytorius Thank you,

Please check your files again. There are a few files that have recorded no change (perhaps end-of-line or end-of-file changes but are listed here as modified. Configure your git to ignore whitespace changes.

Also, it is not a good practice to change your master branch and push that to a derivative branch. Rather create a new branch that matches your task number and then make your changes against this. You can then submit a pull request to merge that branch to the target branch (in this case v1.5.10.1)

chrispytorius commented:
@pradeshc 

Thanks for the help, i have updated the pull request with the whitespace changes removed, and i have updated to the correct target branch. Hopefully this is now correct.

boban-dj commented:
Hi, I was wondering if we could make this expenses module work.
I see there are some issues from these commits, was it tested and working or a work in progress?

rein-10 commented:
i cannot make this work i get this error: 

Error Number: 1146

Table 'transene_invo873.ip_expense_custom' doesn't exist

SELECT * FROM `ip_expense_custom` WHERE `expense_id` IS NULL

Filename: core/MY_Model.php

Line Number: 157

Can someone guide me in the correct direction?

boban-dj commented:
add those missing tabels to the db

On Thu, Apr 23, 2020, 20:29 rein-10 <notifications@github.com> wrote:

> i cannot make this work i get this error:
>
> Error Number: 1146
>
> Table 'transene_invo873.ip_expense_custom' doesn't exist
>
> SELECT * FROM ip_expense_custom WHERE expense_id IS NULL
>
> Filename: core/MY_Model.php
>
> Line Number: 157
>
> Can someone guide me in the correct direction?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/InvoicePlane/InvoicePlane/pull/688#issuecomment-618573489>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABI5UAV4NSQE4Z5FIBXEJJLROCCIBANCNFSM4GX3B6CA>
> .
>

MartinH0 commented:
If anything should be added to the DB this also should be added to [/application/modules/setup/sql/](https://github.com/InvoicePlane/InvoicePlane/tree/master/application/modules/setup/sql) in the related .sql file. In this example you should create **035_1.5.12.sql** and add all the DB changes so they will be executed in the update process.

rein-10 commented:
> add those missing tabels to the db
> […](#)
> On Thu, Apr 23, 2020, 20:29 rein-10 ***@***.***> wrote: i cannot make this work i get this error: Error Number: 1146 Table 'transene_invo873.ip_expense_custom' doesn't exist SELECT * FROM ip_expense_custom WHERE expense_id IS NULL Filename: core/MY_Model.php Line Number: 157 Can someone guide me in the correct direction? — You are receiving this because you commented. Reply to this email directly, view it on GitHub <[#688 (comment)](https://github.com/InvoicePlane/InvoicePlane/pull/688#issuecomment-618573489)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABI5UAV4NSQE4Z5FIBXEJJLROCCIBANCNFSM4GX3B6CA> .

Thanks for the comment. Didnt really know how but finally figured it out. Is it possible to attach multiple files, instead of just one?

boban-dj commented:
from what I saw the file upload is just 1. but its not using existing fileupload from dropzone.js as in invoices and qoutes.

rein-10 commented:
Does anyone else is having trouble with the taxable amount paid? I think the formula is not correct since if I reduce the tax rate, the tax paid increases.
ie: 
Input Tax Rate (16%) = Taxable amount paid (8.62%)
Input Tax Rate (10%) = Tax Rate  (9.09%)

![image](https://user-images.githubusercontent.com/63319512/81429786-f051d880-9123-11ea-9600-dd878c6b36f5.png)

rein-10 commented:
The tax amount paid only works correctly for 10%? If you change the percentage it will show an incorrect calculation. You can fix it by changing the divisor but it would be a problem if you need to manage more than one tax rates. Am I right or is there something that i am missing?

clockwiseq commented:
I'm assuming this is a precision issue.  Sounds like .1 is fine, but .09 is rounding or dropping the precision.

rein-10 commented:
> I'm assuming this is a precision issue. Sounds like .1 is fine, but .09 is rounding or dropping the precision.

Yes any other value instead of .1, will make the calculation imprecise.

rein-10 commented:
Yes any other value instead of .1, will make the calculation imprecise. Made this correction and it will work with all the values that I tried, dont know really much of PHP but this work for me.

partial_expense_table.php 
Line 31
<td><?php echo format_currency($expense->expense_amount / (($expense->tax_rate_percent + 100) / $expense->tax_rate_percent)); ?></td>

mdl_reports.php 
Line 742
$this->db->select('ip_clients.client_name, SUM(ip_expenses.expense_amount) expense_total, SUM(ip_expenses.expense_amount / ((ip_tax_rates.tax_rate_percent + 100) / tax_rate_percent)) expense_tax_total, YEAR(ip_expenses.expense_date) year', FALSE);


 Hope this helps!

arturtle commented:
I keep on getting this error when trying to vie the index. I realize I need to add ip_expenses to my database, what I don't understand is what I should add to my columns. Can someone give me a hand?

> Error Number: 1146
> 
> Table 'alsko_invo429.ip_expenses' doesn't exist
> 
> SELECT SQL_CALC_FOUND_ROWS ip_expense_custom.*, ip_payment_methods.*, ip_clients.client_name, ip_clients.client_id, ip_tax_rates.tax_rate_name, ip_tax_rates.tax_rate_percent, ip_expenses.* FROM `ip_expenses` LEFT JOIN `ip_payment_methods` ON `ip_payment_methods`.`payment_method_id` = `ip_expenses`.`payment_method_id` LEFT JOIN `ip_expense_custom` ON `ip_expense_custom`.`expense_id` = `ip_expenses`.`expense_id` LEFT JOIN `ip_clients` ON `ip_clients`.`client_id` = `ip_expenses`.`client_id` LEFT JOIN `ip_tax_rates` ON `ip_tax_rates`.`tax_rate_id` = `ip_expenses`.`tax_rate_id` ORDER BY `ip_expenses`.`expense_date` DESC LIMIT 15
> 
> Filename: core/MY_Model.php
> 
> Line Number: 220

Also, how do I get expenses

 to show on my nav? And can I rename the dropdown options?
<img width="939" alt="BD error" src="https://user-images.githubusercontent.com/25350028/86307886-e01b1e80-bbe5-11ea-9b45-0498b774e4ca.PNG">
<img width="494" alt="expense_1" src="https://user-images.githubusercontent.com/25350028/86307888-e0b3b500-bbe5-11ea-93c4-57ce04bff9fb.PNG">

pradeshc commented:
@arturtle , You will have to revisit your installation of the module.

The module needs the `ip_expenses` table to run, and from the error message it is clear that the table is not there.



> I keep on getting this error when trying to vie the index. I realize I need to add ip_expenses to my database, what I don't understand is what I should add to my columns. Can someone give me a hand?
> 
> > Error Number: 1146
> > Table 'alsko_invo429.ip_expenses' doesn't exist

BastianBalthasarBux commented:
Is there any progress on this? Further plans?

nielsdrost7 commented:
@BastianBalthasarBux the progress is, that this module is built in V2 of InvoicePlane.
The V1 part, the PR you're looking at right now hasn't had progress since July of 2020.

The V2 progress is coming along quite nicely.
It doesn't have expense_items, like quotes and invoices have, but for the rest: happy with the results in V2

BastianBalthasarBux commented:
@nielsdrost7 this sounds really great!
[i do not need expense items, just "expense" would be great, so I could push the invoices i paid and have some sort of "poor mans" cash accounting ...]

nielsdrost7 commented:
@BastianBalthasarBux understood. I'll add you to slack, we sometimes talk about V2 development there

BastianBalthasarBux commented:
@nielsdrost7 i am already there since ... hmm, ages I think ... ;)
